README
Gets weather from http://foehn.colorado.edu/weather/atoc1/ and displays (some of) it in readable form on your android device.

There are two ways to download this file:
	- the binary in the bin/ directory, which contains a debug .apk. This includes Log statements. This is possibly broken at any given time.
	- the binary on the Downloads page, which has had Log statements stripped out. This is guaranteed stable-ish. (as of this release).


Version: 1.3.2

Todo for future versions:
    Icons
    Clean up the horrible html scraping job I do
    "Graph view" that scrapes the graphs from the page as well.


changelog.txt
v 1.0beta3
	- totally broken build that introduced crash on start (whoops)

v 1.0beta3notfixed
	- semi-broken build
	- moved the crash on start to crash on screen rotation
	- implemented a very poor (see above) implementation of caching.

v 1.0beta4
	- fixed crash on rotating the screen
	- major refactoring, sucks a little bit less now
	- removed caching

V 1.0
	- fixed screen rotation looking silly
	- "final" 1.0 release

v 1.1
	- Added menu functionality
	- Added support for remaining table rows

v 1.2
	- Added simple view and set it to default
	- Altered WeatherMap to more easily support the simple view
	- Added menu option to switch between simple and detailed view

v 1.3
	- Made buttons less horrible looking (spacing!)
	- Addded Mode dialog ("Current Weather", "Minimum Weather", etc)
	- Made it so simple view shows min/max times on the same line (un-squished the column)

v1.3.1
	- Bug fix to landscape layout... I think.

v1.3.2
	- Fixed the additional date field bug.


src/com/emanlaerton/WeatherMap.java
package com.emanlaerton;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

//import android.util.Log;

/*
 * A class to abstract the Map<String,List<String>> currently used and turn it into something a bit more safe
 */
public class WeatherMap {
	private static final String prefix = "WeatherMap";
	private String date;
	private Map<String,String[]> map; //key = name of the type of weather
	public static final Integer CURRENT = 0;
	public static final Integer MINIMUM = 1;
	public static final Integer MAXIMUM = 2;
	public static final Integer AVERAGE = 3;
	
	WeatherMap(String date){
		this.date = date;
		this.map = new LinkedHashMap<String,String[]>();//linked to preserve ordering, I think. But it doesn't really matter as long as it's ordered.
	}
	
	public void setDate(String date) {
		this.date = date;
	}
	
	public String getDate(){
		if (date==null){
			return new String("");//empty string is better than null...
		}
		return this.date;
	}
	
	public Set<String> keys(){
		if(map==null){ //should never happen
			return null;
		}
		return this.map.keySet();
	}
	
	//add a list of values for a weather attribute
	public void insertValues(String key, String[] value){
		
		//go through it and remove all the empty strings
		Integer i = value.length;
		for(i=0; i<value.length; i++){
			if(value[i]==null){
				value[i]="";
			}
			if(value[i].replaceAll("[ \t]","").length()==0){
				value[i]="";
			}
		}
		
		//now put our values in the list
		this.map.put(key, value);
	}
	
	//get the list of values for a weather attribute
	public String[] getValues(String key){
		String[] values = this.map.get(key);
		if(values!=null){
			return values;
		} else {
			//Log.w(prefix,"Warning, no values");
			return null;
		}
	}
	
	public boolean nullDate(){
		return (this.date==null);
	}
	
	//SHOULD NEVER HAPPEN
	public boolean nullMap(){
		return (this.map==null);
	}
	
	
	
}


src/com/emanlaerton/WeatherHelper.java
package com.emanlaerton;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import org.apache.http.HttpResponse;
//import android.util.Log;
/*
 * This function scrapes the page it is passed (sure hope it's the atoc1 page) and converts it into a WeatherMap.
 */
public class WeatherHelper {
	//private final static String prefix = "WeatherHelper";
	
	public static WeatherMap read(HttpResponse response){
		return parseAtoc(request(response));
	}
	
    private static String request(HttpResponse response){
        String result = "";
        try{
            InputStream in = response.getEntity().getContent();
            BufferedReader reader = new BufferedReader(new InputStreamReader(in));
            StringBuilder str = new StringBuilder();
            String line = null;
            while((line = reader.readLine()) != null){
                str.append(line + "\n");
            }
            in.close();
            result = str.toString();
        }catch(Exception ex){
        	result = "Error";
        }
        return result;
	}
    
    //Ugh.
    private static WeatherMap parseAtoc(String pageHTML){
    	String date="";
    	String labelLine = "<tr bgcolor=#> <th align=left>"; 
    	String currentLine = "</th> <td align=center>";
    	String minLine = "</td> <td align=right>";
    	String maxLine = "</td> <td align=right>";
    	String avgLine = "</td> <td align=center>";
    	String endLine = "</td> </tr>";
    	String weirdLine = "</td> <td>"; //this handles problem lines
    	String delims[] = {currentLine,minLine,maxLine,avgLine,endLine};
    	String Label;
    	String values[];
    	WeatherMap wm = null;
    	try {
    		int startCut = pageHTML.indexOf("<table border");
    		int endCut = pageHTML.indexOf("</table>");
    		//cut to table
    		pageHTML = pageHTML.substring(startCut, endCut);
    		
    		//retrieve the date
    		startCut = pageHTML.indexOf("<tr> <th align=center>");
    		endCut = pageHTML.indexOf("</th>");
    		date = pageHTML.substring(startCut,endCut).replaceFirst("<tr> <th align=center>", "").trim();
    		
    		//clean page of annoying differences in table
    		pageHTML = pageHTML.replaceAll("ffff[48][48]", "");
    		//fix the stupid <sup></sup>
    		pageHTML = pageHTML.replaceAll("m<sup>2</sup>","m^2");
    		
    		//cut to next line
    		startCut = pageHTML.indexOf(labelLine);
    		pageHTML = pageHTML.substring(startCut+labelLine.length());
    		
    		wm = new WeatherMap(date);
    		
    		//we want 9 rows
    		for(int i=0;i<9;i++){
    			values = new String[4];
    			Integer loc=0;
    			//endCut is the character at the start of the first instance of "</th> <td align=center>"
    			endCut = pageHTML.indexOf(delims[0]);
    			//Label is from the start of the the align left statement to the end cut - the label!
    			Label = pageHTML.substring(0, endCut).trim();
    			pageHTML = pageHTML.substring(endCut+delims[0].length());
    			//delims.length=4, or so we hope (cur, min, max, avg)
    			for(int j=0;j<(delims.length-1);j++){
    				endCut = pageHTML.indexOf(delims[j+1]);
    				if(endCut==-1){ //uh-oh, it wasn't there... was it a non-standard row with an extra endline?
    					endCut = pageHTML.indexOf(weirdLine);
    					if(endCut==-1){ //give up and don't do this row any more, and don't add anything to the map
    						continue;
    					} else { //ok, what we've got to the map
    						break;
    					}
    				}
    				values[loc++] = pageHTML.substring(0,endCut).replaceAll("[ \t]",""); 
    				pageHTML = pageHTML.substring(endCut+delims[j+1].length());
    			}
        		startCut = pageHTML.indexOf(labelLine);
        		pageHTML = pageHTML.substring(startCut+labelLine.length());
    			wm.insertValues(Label, values);
    		}
    		
		} catch (Exception e) {
			pageHTML = "Error, things went wrong somewhere in parsing";
			e.printStackTrace();
		}
		////Log.i(prefix,"Result: "+pageHTML);
    	return wm;
    }
} 


src/com/emanlaerton/BoulderWeather.java
package com.emanlaerton;

import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.DefaultHttpClient;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
//import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ListView;
import android.widget.SimpleAdapter;
import android.widget.TextView;
import android.widget.Toast;
import android.os.SystemClock;


public class BoulderWeather extends Activity {
	private final static long MINUTE = 1000*60;
	private final static long REFRESH_TIMEOUT = 5 * MINUTE; //sounds about right
	private final static String ATOC_URL = "http://foehn.colorado.edu/weather/atoc1/"; 
	TextView date = null;
	TextView viewText = null;
	private static long lastRefresh=0;
	private static final String prefix = "BoulderWeather";
	private volatile static WeatherMap weatherMap = null; //A Map of the weather, get it? I kill me.
	private Integer view = WeatherMap.CURRENT; //default
    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        
        
        date = (TextView) findViewById(R.id.weather_date);
        setContentView(R.layout.main);
        Button refresh = (Button) findViewById(R.id.refresh_button);
        Button svone = (Button) findViewById(R.id.svone);
        Button svtwo = (Button) findViewById(R.id.svtwo);
        Button svthree = (Button) findViewById(R.id.svthree);
        viewText = (TextView) findViewById(R.id.weather_mode);
        viewText.setText(getView(view)+" Weather");
	    refresh.setOnClickListener(new View.OnClickListener(){
			public void onClick(View v) {
				lastRefresh=SystemClock.elapsedRealtime();
				weatherMap = getWeather(ATOC_URL);
				drawScreen(weatherMap);
			}
	    });
	    svone.setOnClickListener(new View.OnClickListener(){
	    	public void onClick(View v) {
	    		switchView(1);
	    		refreshIfTimerElapsed();
	    	}
	    });
	    svtwo.setOnClickListener(new View.OnClickListener(){
	    	public void onClick(View v) {
	    		switchView(2);
	    		refreshIfTimerElapsed();
	    	}
	    });
	    svthree.setOnClickListener(new View.OnClickListener(){
	    	public void onClick(View v) {
	    		switchView(3);
	    		refreshIfTimerElapsed();
	    	}
	    });
    }
    /** Called when the activity is started **/
    @Override
    public void onStart(){
    	super.onStart();
    	if(date!=null){ //just in case
    		return;
    	}
    	refreshIfTimerElapsed();
    	
    }
    
    public void refreshIfTimerElapsed(){
    	long currentTime=SystemClock.elapsedRealtime();
    	if (lastRefresh==0){
    		lastRefresh=currentTime;
    		weatherMap = getWeather(ATOC_URL);
    		drawScreen(weatherMap);
    		//Log.i(prefix,"lastRefresh=0, just opened");
    	} else if((currentTime-lastRefresh) > REFRESH_TIMEOUT){
    		//Log.i(prefix,"currentTime="+currentTime+" lastRefresh="+lastRefresh+" difference="+(currentTime-lastRefresh));
    		lastRefresh=currentTime;
    		weatherMap = getWeather(ATOC_URL);
    		drawScreen(weatherMap);
    	} //else don't refresh again, but do the screen draw
    	else {
    		//Log.i(prefix,"else statement hit!");
    		//if we've got the weather stored, use that. otherwise, we'll have to get it again.
    		if(weatherMap==null){ //this indicates Dalvik went and destroyed my activity. Time to rebuild the map.
    			//Log.i(prefix,"Getting the weather even though it should be stored");
    			lastRefresh=currentTime;
    			weatherMap = getWeather(ATOC_URL);
    		}
    		drawScreen(weatherMap);
    	}
    }
    
    /* make a WeatherHelper and use it request our nice map */
    private WeatherMap getWeather(String url){
    	WeatherMap weatherMap = null; //local weatherMap while we build
		HttpClient client = new DefaultHttpClient();
		HttpGet request = new HttpGet(url);
		try{
			HttpResponse response = client.execute(request);
			weatherMap = WeatherHelper.read(response);
		}catch(UnknownHostException ex){
			ex.printStackTrace();
			//Log.w(prefix,"Probably couldn't connect to the internet");
			Toast.makeText(getApplicationContext(),"Could not connect! Is your internet connection down?", Toast.LENGTH_SHORT).show();
			return null;
    	}catch(Exception ex){
			ex.printStackTrace();
			return null;
		}
		
		return weatherMap;
    }
    
    /* Draw a screen based on a nice map. If it's not a nice map, probably crash in some unpleasant and unexpected way. 
     */
    private boolean drawScreen(WeatherMap wm){
        //final TextView 
        final TextView date = (TextView) findViewById(R.id.weather_date);
		if(wm==null){
			//Log.w(prefix,"Warning, Map was not returned properly!");
			Toast.makeText(getApplicationContext(),"Weather data could not be found!", Toast.LENGTH_SHORT).show();
			date.setText("No results! Check your internet connection.");
		}else{
			ListView lv= (ListView)findViewById(R.id.listview);
	        String[] from = new String[] { "Type", getView(view) };
	        int[] to = new int[] { R.id.Key, R.id.Value};
	        List<HashMap<String, String>> filledMaps = new ArrayList<HashMap<String, String>>();
	        String ps = getResources().getStringArray(R.array.prepStrings)[view];
	        //prep the list
        	//Log.i(prefix, "view: "+getView(view));
	        for(String s : wm.keys()){
				//Log.i(prefix,"Stat: "+s);
				//Log.i(prefix, "Value: "+wm.getValues(s)[view]);
	        	
	        	
	    		HashMap<String, String> prepMap = new HashMap<String,String>();
	    		prepMap.put("Type", s);
	    		String[] rowString = wm.getValues(s);
	    		prepMap.put(ps,rowString[view]);
	    		filledMaps.add(prepMap);
	        }
	        SimpleAdapter adapter = new SimpleAdapter(this, filledMaps, R.layout.simple, from, to);
	        lv.setAdapter(adapter);
			date.setText("Date: "+wm.getDate());
		}
    	return true;
    }
    
    private String getView(Integer view) {
    	String ps[] = getResources().getStringArray(R.array.prepStrings);
    	if(view < ps.length){
    		return ps[view];
    	}
    	
		return null;
	}
    private boolean switchView(Integer button){
    	Integer newView;
    	String ps[] = getResources().getStringArray(R.array.prepStrings);
    	//View is 0: Button n -> view n
    	//View is 1: Button 1 -> view 0, 2->2, 3->3 ((v == b) ? b-1 : b) 
    	//View is 2: 1->0, 2->1, 3->3 ((v <= b
    	//View is 3: 1->0, 2->1, 3->2
    	newView = ((button > view) ? button : button-1);
    	Button sv[] = new Button[3];
        sv[0] = (Button) findViewById(R.id.svone);
        sv[1] = (Button) findViewById(R.id.svtwo);
        sv[2] = (Button) findViewById(R.id.svthree);
        for(Integer i=0; i<3; i++){
        	String text = (i < newView) ? ps[i] : ps[i+1];
        	sv[i].setText(text);
        }
        view = newView;
        if(viewText!=null){
        	viewText.setText(getView(view)+" Weather");
        }
    	
    	return true;
    }
	/* Make menu stuff happen */
	 @Override
	    public boolean onCreateOptionsMenu(Menu menu) {
	    	//Load Menu
	       	MenuInflater inflater = getMenuInflater();
	        inflater.inflate(R.menu.menu, menu);
	        return true;
	    }
	 @Override
	    public boolean onOptionsItemSelected(MenuItem item) {
	        switch (item.getItemId()) {
            case R.id.Quit:
            	//Log.w(prefix,"Moving to back");
            	//we shouldn't quit, so instead we'll fake it
            	this.moveTaskToBack(true);
            	break;
            case R.id.Refresh:
            	//Log.w(prefix,"Refreshing...");
				lastRefresh=SystemClock.elapsedRealtime();
				weatherMap = getWeather(ATOC_URL);
				drawScreen(weatherMap);
				break;
            case R.id.Switch:
            	//Log.w(prefix, "Switching to details view");
            	Intent i = new Intent(this, BWDetails.class);
            	this.startActivity(i);
            	break;
        }
	        return true;
	    }
    
}

src/com/emanlaerton/BWDetails.java
package com.emanlaerton;

import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.DefaultHttpClient;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.SystemClock;
//import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ListView;
import android.widget.SimpleAdapter;
import android.widget.TextView;
import android.widget.Toast;


public class BWDetails extends Activity {
	private final static long MINUTE = 1000*60;
	private final static long REFRESH_TIMEOUT = 5 * MINUTE; //sounds about right
	private final static String ATOC_URL = "http://foehn.colorado.edu/weather/atoc1/"; 

	TextView date = null;
	private static long lastRefresh=0;
	private static final String prefix = "BoulderWeather";
	private volatile static WeatherMap weatherMap = null; //A Map of the weather, get it? I kill me.
    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        
        
        date = (TextView) findViewById(R.id.weather_date);
        setContentView(R.layout.details);
        Button refresh = (Button) findViewById(R.id.refresh_button);
	    refresh.setOnClickListener(new View.OnClickListener(){
			public void onClick(View v) {
				lastRefresh=SystemClock.elapsedRealtime();
				weatherMap = getWeather(ATOC_URL);
				drawScreen(weatherMap);
			}
	    	
	    });
    }
    /** Called when the activity is started **/
    @Override
    public void onStart(){
    	super.onStart();
    	if(date!=null){ //just in case
    		return;
    	}
    	refreshIfTimerElapsed();
    	
    }
    
    public void refreshIfTimerElapsed(){
    	long currentTime=SystemClock.elapsedRealtime();
    	if (lastRefresh==0){
    		lastRefresh=currentTime;
    		weatherMap = getWeather(ATOC_URL);
    		drawScreen(weatherMap);
    		//Log.i(prefix,"lastRefresh=0, just opened");
    	} else if((currentTime-lastRefresh) > REFRESH_TIMEOUT){
    		//Log.i(prefix,"currentTime="+currentTime+" lastRefresh="+lastRefresh+" difference="+(currentTime-lastRefresh));
    		lastRefresh=currentTime;
    		weatherMap = getWeather(ATOC_URL);
    		drawScreen(weatherMap);
    	} //else don't refresh again, but do the screen draw
    	else {
    		//Log.i(prefix,"else statement hit!");
    		//if we've got the weather stored, use that. otherwise, we'll have to get it again.
    		if(weatherMap==null){ //this indicates Dalvik went and destroyed my activity. Time to rebuild the map.
    			//Log.i(prefix,"Getting the weather even though it should be stored");
    			lastRefresh=currentTime;
    			weatherMap = getWeather(ATOC_URL);
    		}
    		drawScreen(weatherMap);
    	}
    }
    
    /* make a WeatherHelper and use it request our nice map */
    private WeatherMap getWeather(String url){
    	WeatherMap weatherMap = null; //local weatherMap while we build
		HttpClient client = new DefaultHttpClient();
		HttpGet request = new HttpGet(url);
		try{
			HttpResponse response = client.execute(request);
			weatherMap = WeatherHelper.read(response);
		}catch(UnknownHostException ex){
			ex.printStackTrace();
			//Log.w(prefix,"Probably couldn't connect to the internet");
			Toast.makeText(getApplicationContext(),"Could not connect! Is your internet connection down?", Toast.LENGTH_SHORT).show();
			return null;
    	}catch(Exception ex){
			ex.printStackTrace();
			return null;
		}
		
		return weatherMap;
    }
    
    /* Draw a screen based on a nice map. If it's not a nice map, probably crash in some unpleasant and unexpected way. 
     */
    private boolean drawScreen(WeatherMap wm){
        //final TextView 
        final TextView date = (TextView) findViewById(R.id.weather_date);
		if(wm==null){
			//Log.w(prefix,"Warning, Map was not returned properly!");
			Toast.makeText(getApplicationContext(),"Weather data could not be found!", Toast.LENGTH_SHORT).show();
			date.setText("No results! Check your internet connection.");
		}else{
			ListView lv= (ListView)findViewById(R.id.listview);
	        String[] from = new String[] { "Type", "Current", "Minimum", "Maximum", "Average" };
	        int[] to = new int[] { R.id.Value, R.id.Current_val, R.id.Minimum_val, R.id.Maximum_val, R.id.Average_val };
	        List<HashMap<String, String>> filledMaps = new ArrayList<HashMap<String, String>>();
	        String [] ps = getResources().getStringArray(R.array.prepStrings);
	        //prep the list
	        for(String s: wm.keys()){
	        	//Log.i(prefix,"Stat: "+s);
	        	//Log.i(prefix, "Value: "+wm.getValues(s));
	        	
	        	//This is pretty brittle
	    		HashMap<String, String> prepMap = new HashMap<String,String>();
	    		prepMap.put("Type", s);
	    		String[] rowString = wm.getValues(s);
	    		for(int i=0;i<wm.getValues(s).length;i++){
	    			prepMap.put(ps[i],rowString[i]);
	    		}
	    		filledMaps.add(prepMap);
	        }
	        SimpleAdapter adapter = new SimpleAdapter(this, filledMaps, R.layout.results, from, to);
	        lv.setAdapter(adapter);
			date.setText("Date: "+wm.getDate());
		}
    	return true;
    }
    
    /* Make menu stuff happen */
	 @Override
	    public boolean onCreateOptionsMenu(Menu menu) {
	    	//Load Menu
	       	MenuInflater inflater = getMenuInflater();
	        inflater.inflate(R.menu.menu, menu);
	        return true;
	    }
	 @Override
	    public boolean onOptionsItemSelected(MenuItem item) {
	        switch (item.getItemId()) {
	            case R.id.Quit:
	            	//Log.w(prefix,"Moving to back");
	            	//we shouldn't quit, so instead we'll fake it
	            	this.moveTaskToBack(true);
	            	break;
	            case R.id.Refresh:
	            	//Log.w(prefix,"Refreshing...");
					lastRefresh=SystemClock.elapsedRealtime();
					weatherMap = getWeather(ATOC_URL);
					drawScreen(weatherMap);
					break;
	            case R.id.Switch:
	            	//Log.w(prefix, "Switching to simple view");
	            	Intent i = new Intent(this, BoulderWeather.class);
	            	this.startActivity(i);
	            	break;
	        }
	        return true;
	    }
    
}